circuit/src/native.rs

```diff
@@ -0,0 +1,403 @@
+			Fr::zero(),
+		];
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(&sk1, &pk1, message_hashes[0]);
+
+		let scores = pks.zip(scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+
+		set.update_op(pk1, op);
+
+		// Peer2(pk2) signs the opinion
+		let pks = [
+			pk1,
+			pk2,
+			PublicKey::default(),
+			PublicKey::default(),
+			PublicKey::default(),
+			PublicKey::default(),
+		];
+		let scores = [
+			Fr::from_u128(INITIAL_SCORE),
+			Fr::zero(),
+			Fr::zero(),
+			Fr::zero(),
+			Fr::zero(),
+			Fr::zero(),
+		];
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(&sk2, &pk2, message_hashes[0]);
+
+		let scores = pks.zip(scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+
+		set.update_op(pk2, op);
+
+		set.converge();
+	}
+
+	#[test]
+	fn test_add_three_members_with_opinions() {
+		let mut set = EigenTrustSet::new();
+
+		let rng = &mut thread_rng();
+
+		let sk1 = SecretKey::random(rng);
+		let sk2 = SecretKey::random(rng);
+		let sk3 = SecretKey::random(rng);
+
+		let pk1 = sk1.public();
+		let pk2 = sk2.public();
+		let pk3 = sk3.public();
+
+		set.add_member(pk1);
+		set.add_member(pk2);
+		set.add_member(pk3);
+
+		// Peer1(pk1) signs the opinion
+		let pks = [pk1, pk2, pk3, PublicKey::default(), PublicKey::default(), PublicKey::default()];
+		let scores = [Fr::zero(), Fr::from(300), Fr::from(700), Fr::zero(), Fr::zero(), Fr::zero()];
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(&sk1, &pk1, message_hashes[0]);
+
+		let scores = pks.zip(scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+
+		set.update_op(pk1, op);
+
+		// Peer2(pk2) signs the opinion
+		let pks = [pk1, pk2, pk3, PublicKey::default(), PublicKey::default(), PublicKey::default()];
+		let scores = [Fr::from(600), Fr::zero(), Fr::from(400), Fr::zero(), Fr::zero(), Fr::zero()];
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(&sk2, &pk2, message_hashes[0]);
+
+		let scores = pks.zip(scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+
+		set.update_op(pk2, op);
+
+		// Peer3(pk3) signs the opinion
+		let pks = [pk1, pk2, pk3, PublicKey::default(), PublicKey::default(), PublicKey::default()];
+		let scores = [Fr::from(600), Fr::from(400), Fr::zero(), Fr::zero(), Fr::zero(), Fr::zero()];
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(&sk3, &pk3, message_hashes[0]);
+
+		let scores = pks.zip(scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+
+		set.update_op(pk3, op);
+
+		set.converge();
+	}
+
+	#[test]
+	#[should_panic]
+	fn test_add_three_members_with_two_opinions() {
```

Comment:
We should have two versions of this test
1) We should have 2 peers giving scores to each others, and the third giving scores to invalid peers, so the third one should be invalidate and the scores should successfully converge
2) Peer 1 should give score to Peer 2, Peer 2 should give score to Peer 3, and Peer 3 should give scores to invalid peers. Peer 3 will get invalidated, which means Peer 2 will get invalidated, which also means Peer 1 will get invalidated. The scores should not be able to converge

Reply:
I think the first test is done in `test_add_three_members_with_two_opinions()` test example.

For the second test, I suspect if it needs more complicated algorithm(related to graph?).
It seems that I should check the connectivity between peers, and it needs much more than forward/backward loop.